### ISSUE TITLE
mutex lock failed: Invalid argument when importing after tink (M1, macOS 14.2.1)

### ISSUE THREAD

**Issue by reeseann** [apache/arrow#40088]
### Describe the bug, including details regarding any error messages, version, and platform.

- macOS 14.2.1, M1
- clang 15.0.0 (clang-1500.1.0.2.5, arm64-apple-darwin23.2.0)
- Python 3.8.18
- Pyarrow 15.0.0
- Tink 1.9.0 (install used prebuilt wheel tink-1.9.0-cp38-cp38-macosx_11_0_universal2.whl)
- OpenSSL 3.2.0 23 Nov 2023

```python
import tink
import pyarrow
```

Results in:

```
libc++abi: terminating due to uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
```

Reversing import order results in the process hanging with:

```
[mutex.cc : 453] RAW: Lock blocking 0x156747c38   @
```

Same error message as #15189. Just in case similar root cause, OpenSSL version Python is built with is `OpenSSL 3.2.0 23 Nov 2023`.

### Component(s)

Python

**Comment by pitrou** [comment]
Can you perhaps obtain a backtrace using lldb?

**Comment by reeseann** [comment]
```
0.	Program arguments: /Library/Developer/CommandLineTools/usr/bin/lldb /Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/bin/python3
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  lldb                     0x000000010414f7dc llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 56
1  lldb                     0x000000010414ed38 llvm::sys::RunSignalHandlers() + 112
2  lldb                     0x000000010414fe14 SignalHandler(int) + 304
3  libsystem_platform.dylib 0x000000018180da24 _sigtramp + 56
4  libsystem_pthread.dylib  0x00000001817ddcc0 pthread_kill + 288
5  libsystem_c.dylib        0x00000001816e9a40 abort + 180
6  libc++abi.dylib          0x0000000181795070 __cxxabiv1::__aligned_malloc_with_fallback(unsigned long) + 0
7  libc++abi.dylib          0x0000000181785110 demangling_terminate_handler() + 320
8  libobjc.A.dylib          0x000000018142b99c _objc_terminate() + 160
9  libc++abi.dylib          0x0000000181794434 std::__terminate(void (*)()) + 16
10 libc++abi.dylib          0x0000000181797520 __cxa_get_exception_ptr + 0
11 libc++abi.dylib          0x0000000181797464 __cxxabiv1::exception_cleanup_func(_Unwind_Reason_Code, _Unwind_Exception*) + 0
12 libc++.1.dylib           0x00000001817136b8 std::__1::__throw_system_error(int, char const*) + 100
13 libc++.1.dylib           0x0000000181708318 std::__1::mutex::try_lock() + 0
14 libarrow.1500.dylib      0x0000000110d6a330 google::protobuf::internal::OnShutdownRun(void (*)(void const*), void const*) + 72
15 libarrow.1500.dylib      0x0000000110d61880 google::protobuf::internal::InitProtobufDefaultsSlow() + 84
16 libarrow.1500.dylib      0x0000000110dfaca8 google::protobuf::(anonymous namespace)::AddDescriptors(google::protobuf::internal::DescriptorTable const*) + 128
17 libarrow.1500.dylib      0x0000000110dfacec google::protobuf::internal::AddDescriptorsRunner::AddDescriptorsRunner(google::protobuf::internal::DescriptorTable const*) + 24
18 dyld                     0x0000000181479a24 invocation function for block in dyld4::Loader::findAndRunAllInitializers(dyld4::RuntimeState&) const::$_0::operator()() const + 168
19 dyld                     0x00000001814bf0f4 invocation function for block in dyld3::MachOAnalyzer::forEachInitializer(Diagnostics&, dyld3::MachOAnalyzer::VMAddrConverter const&, void (unsigned int) block_pointer, void const*) const + 172
20 dyld                     0x00000001814b2668 invocation function for block in dyld3::MachOFile::forEachSection(void (dyld3::MachOFile::SectionInfo const&, bool, bool&) block_pointer) const + 496
21 dyld                     0x00000001814592fc dyld3::MachOFile::forEachLoadCommand(Diagnostics&, void (load_command const*, bool&) block_pointer) const + 300
22 dyld                     0x00000001814b16a0 dyld3::MachOFile::forEachSection(void (dyld3::MachOFile::SectionInfo const&, bool, bool&) block_pointer) const + 192
23 dyld                     0x00000001814b4188 dyld3::MachOFile::forEachInitializerPointerSection(Diagnostics&, void (unsigned int, unsigned int, bool&) block_pointer) const + 160
24 dyld                     0x00000001814bede8 dyld3::MachOAnalyzer::forEachInitializer(Diagnostics&, dyld3::MachOAnalyzer::VMAddrConverter const&, void (unsigned int) block_pointer, void const*) const + 432
25 dyld                     0x0000000181475b38 dyld4::Loader::findAndRunAllInitializers(dyld4::RuntimeState&) const + 524
26 dyld                     0x000000018147bf70 dyld4::JustInTimeLoader::runInitializers(dyld4::RuntimeState&) const + 36
27 dyld                     0x0000000181475f24 dyld4::Loader::runInitializersBottomUp(dyld4::RuntimeState&, dyld3::Array<dyld4::Loader const*>&) const + 220
28 dyld                     0x0000000181475f00 dyld4::Loader::runInitializersBottomUp(dyld4::RuntimeState&, dyld3::Array<dyld4::Loader const*>&) const + 184
29 dyld                     0x0000000181475f00 dyld4::Loader::runInitializersBottomUp(dyld4::RuntimeState&, dyld3::Array<dyld4::Loader const*>&) const + 184
30 dyld                     0x0000000181475f00 dyld4::Loader::runInitializersBottomUp(dyld4::RuntimeState&, dyld3::Array<dyld4::Loader const*>&) const + 184
31 dyld                     0x0000000181475f00 dyld4::Loader::runInitializersBottomUp(dyld4::RuntimeState&, dyld3::Array<dyld4::Loader const*>&) const + 184
32 dyld                     0x0000000181479ab0 dyld4::Loader::runInitializersBottomUpPlusUpwardLinks(dyld4::RuntimeState&) const::$_1::operator()() const + 112
33 dyld                     0x00000001814760f0 dyld4::Loader::runInitializersBottomUpPlusUpwardLinks(dyld4::RuntimeState&) const + 380
34 dyld                     0x00000001814947ac dyld4::APIs::dlopen_from(char const*, int, void*) + 1576
35 Python3                  0x0000000104ccedf4 _Py_GetLocaleconvNumeric + 756
36 Python3                  0x0000000104c9f66c PyImport_AppendInittab + 5100
37 Python3                  0x0000000104c9ee08 PyImport_AppendInittab + 2952
38 Python3                  0x0000000104beb188 PyCMethod_New + 764
39 Python3                  0x0000000104c77e50 _PyEval_EvalFrameDefault + 23428
40 Python3                  0x0000000104c7b7ec _PyEval_EvalFrameDefault + 38176
41 Python3                  0x0000000104ba8f10 _PyFunction_Vectorcall + 236
42 Python3                  0x0000000104c7aa38 _PyEval_EvalFrameDefault + 34668
43 Python3                  0x0000000104c77b78 _PyEval_EvalFrameDefault + 22700
44 Python3                  0x0000000104ba8fc8 _PyFunction_Vectorcall + 420
45 Python3                  0x0000000104c7aa38 _PyEval_EvalFrameDefault + 34668
46 Python3                  0x0000000104c77b54 _PyEval_EvalFrameDefault + 22664
47 Python3                  0x0000000104ba8fc8 _PyFunction_Vectorcall + 420
48 Python3                  0x0000000104c7aa38 _PyEval_EvalFrameDefault + 34668
49 Python3                  0x0000000104c77bf4 _PyEval_EvalFrameDefault + 22824
50 Python3                  0x0000000104ba8fc8 _PyFunction_Vectorcall + 420
51 Python3                  0x0000000104c7aa38 _PyEval_EvalFrameDefault + 34668
52 Python3                  0x0000000104c77bf4 _PyEval_EvalFrameDefault + 22824
53 Python3                  0x0000000104ba8fc8 _PyFunction_Vectorcall + 420
54 Python3                  0x0000000104c7aa38 _PyEval_EvalFrameDefault + 34668
55 Python3                  0x0000000104c77bf4 _PyEval_EvalFrameDefault + 22824
56 Python3                  0x0000000104ba8fc8 _PyFunction_Vectorcall + 420
57 Python3                  0x0000000104baa0cc PyObject_CallMethodObjArgs + 492
58 Python3                  0x0000000104baa258 _PyObject_CallMethodIdObjArgs + 112
59 Python3                  0x0000000104c9daf4 PyImport_ImportModuleLevelObject + 1284
60 Python3                  0x0000000104c76888 _PyEval_EvalFrameDefault + 17852
61 Python3                  0x0000000104c7b7ec _PyEval_EvalFrameDefault + 38176
62 Python3                  0x0000000104c72204 PyEval_EvalCode + 80
63 Python3                  0x0000000104c6eb5c _PyAST_ExprAsUnicode + 19156
64 Python3                  0x0000000104beb188 PyCMethod_New + 764
65 Python3                  0x0000000104c77e50 _PyEval_EvalFrameDefault + 23428
66 Python3                  0x0000000104c7b7ec _PyEval_EvalFrameDefault + 38176
67 Python3                  0x0000000104ba8f10 _PyFunction_Vectorcall + 236
68 Python3                  0x0000000104c7aa38 _PyEval_EvalFrameDefault + 34668
69 Python3                  0x0000000104c77b78 _PyEval_EvalFrameDefault + 22700
70 Python3                  0x0000000104ba8fc8 _PyFunction_Vectorcall + 420
71 Python3                  0x0000000104c7aa38 _PyEval_EvalFrameDefault + 34668
72 Python3                  0x0000000104c77b54 _PyEval_EvalFrameDefault + 22664
73 Python3                  0x0000000104ba8fc8 _PyFunction_Vectorcall + 420
74 Python3                  0x0000000104c7aa38 _PyEval_EvalFrameDefault + 34668
75 Python3                  0x0000000104c77bf4 _PyEval_EvalFrameDefault + 22824
76 Python3                  0x0000000104ba8fc8 _PyFunction_Vectorcall + 420
77 Python3                  0x0000000104c7aa38 _PyEval_EvalFrameDefault + 34668
78 Python3                  0x0000000104c77bf4 _PyEval_EvalFrameDefault + 22824
79 Python3                  0x0000000104ba8fc8 _PyFunction_Vectorcall + 420
80 Python3                  0x0000000104baa0cc PyObject_CallMethodObjArgs + 492
81 Python3                  0x0000000104baa258 _PyObject_CallMethodIdObjArgs + 112
82 Python3                  0x0000000104c9daf4 PyImport_ImportModuleLevelObject + 1284
83 Python3                  0x0000000104c76888 _PyEval_EvalFrameDefault + 17852
84 Python3                  0x0000000104c7b7ec _PyEval_EvalFrameDefault + 38176
85 Python3                  0x0000000104c72204 PyEval_EvalCode + 80
86 Python3                  0x0000000104cb62f0 PyParser_ASTFromStringObject + 500
87 Python3                  0x0000000104cb4eb4 PyRun_StringFlags + 152
88 LLDB                     0x0000000116b8f3f4 lldb_private::python::runStringMultiLine(llvm::Twine const&, lldb_private::python::PythonDictionary const&, lldb_private::python::PythonDictionary const&) + 148
89 LLDB                     0x0000000116b9503c lldb_private::ScriptInterpreterPythonImpl::ExecuteMultipleLines(char const*, lldb_private::ExecuteScriptOptions const&) + 884
90 LLDB                     0x0000000116b9a5c8 lldb_private::ScriptInterpreterPythonImpl::LoadScriptingModule(char const*, lldb_private::LoadScriptOptions const&, lldb_private::Status&, std::__1::shared_ptr<lldb_private::StructuredData::Object>*, lldb_private::FileSpec) + 1924
91 LLDB                     0x0000000116d9342c CommandObjectCommandsScriptImport::DoExecute(lldb_private::Args&, lldb_private::CommandReturnObject&) + 328
92 LLDB                     0x00000001167a5ad0 lldb_private::CommandObjectParsed::Execute(char const*, lldb_private::CommandReturnObject&) + 656
93 LLDB                     0x000000011679c75c lldb_private::CommandInterpreter::HandleCommand(char const*, lldb_private::LazyBool, lldb_private::CommandReturnObject&, bool) + 2024
94 LLDB                     0x000000011679ff9c lldb_private::CommandInterpreter::IOHandlerInputComplete(lldb_private::IOHandler&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&) + 828
95 LLDB                     0x00000001166d77c8 lldb_private::IOHandlerEditline::Run() + 304
96 LLDB                     0x00000001166bc0c4 lldb_private::Debugger::RunIOHandlers() + 140
97 LLDB                     0x00000001167a1178 lldb_private::CommandInterpreter::RunCommandInterpreter(lldb_private::CommandInterpreterRunOptions&) + 196
98 LLDB                     0x000000011651d798 lldb::SBDebugger::RunCommandInterpreter(bool, bool) + 124
99 lldb                     0x00000001041402b8 Driver::MainLoop() + 2712
100 lldb                     0x0000000104140d14 main + 2036
101 dyld                     0x000000018145d0e0 start + 2360
```

**Comment by kou** [comment]
It seems that this is happen at https://github.com/protocolbuffers/protobuf/blob/976a6eb6a4f462c6bb0b2a87fcd95f0aeb17eedb/src/google/protobuf/message_lite.cc#L726 .

Is tink statically linked with its bundled Protobuf? (pyarrow does.)  If tink is also linked with its bundled Protobuf, they may be conflicted.

**Comment by Inokinoki** [comment]
I started an investigation:

https://github.com/tink-crypto/tink-py/issues/25#issuecomment-2857960794

My proposition is to upgrade the `protobuf` default version of `pyarrow` to a later one (in the CI), other than `3.21.X`.

**Comment by pitrou** [comment]
@Inokinoki I understand the explanation you give there, but it seems unexpected to me that there would be confusion between two private-linked static libraries. Especially as Python [uses the default `RTLD_LOCAL`](https://github.com/python/cpython/blob/3dfed230928de0f649061782a36691066a0ef058/Include/internal/pycore_import.h#L34-L45) when loading extension modules, so there shouldn't be any name clashes when loading two different extension modules.

**Comment by pitrou** [comment]
I'm just seeing your stack trace btw:
```
87 Python3                  0x0000000104cb4eb4 PyRun_StringFlags + 152
88 LLDB                     0x0000000116b8f3f4 lldb_private::python::runStringMultiLine(llvm::Twine const&, lldb_private::python::PythonDictionary const&, lldb_private::python::PythonDictionary const&) + 148
89 LLDB                     0x0000000116b9503c lldb_private::ScriptInterpreterPythonImpl::ExecuteMultipleLines(char const*, lldb_private::ExecuteScriptOptions const&) + 884
90 LLDB                     0x0000000116b9a5c8 lldb_private::ScriptInterpreterPythonImpl::LoadScriptingModule(char const*, lldb_private::LoadScriptOptions const&, lldb_private::Status&, std::__1::shared_ptr<lldb_private::StructuredData::Object>*, lldb_private::FileSpec) + 1924
91 LLDB                     0x0000000116d9342c CommandObjectCommandsScriptImport::DoExecute(lldb_private::Args&, lldb_private::CommandReturnObject&) + 328
92 LLDB                     0x00000001167a5ad0 lldb_private::CommandObjectParsed::Execute(char const*, lldb_private::CommandReturnObject&) + 656
93 LLDB                     0x000000011679c75c lldb_private::CommandInterpreter::HandleCommand(char const*, lldb_private::LazyBool, lldb_private::CommandReturnObject&, bool) + 2024
94 LLDB                     0x000000011679ff9c lldb_private::CommandInterpreter::IOHandlerInputComplete(lldb_private::IOHandler&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&) + 828
95 LLDB                     0x00000001166d77c8 lldb_private::IOHandlerEditline::Run() + 304
96 LLDB                     0x00000001166bc0c4 lldb_private::Debugger::RunIOHandlers() + 140
97 LLDB                     0x00000001167a1178 lldb_private::CommandInterpreter::RunCommandInterpreter(lldb_private::CommandInterpreterRunOptions&) + 196
98 LLDB                     0x000000011651d798 lldb::SBDebugger::RunCommandInterpreter(bool, bool) + 124
99 lldb                     0x00000001041402b8 Driver::MainLoop() + 2712
100 lldb                     0x0000000104140d14 main + 2036
101 dyld                     0x000000018145d0e0 start + 2360
```

Is this the [Python scripting facility for LLDB](https://lldb.llvm.org/use/python.html)? Is it where the crash is happening?

**Comment by Inokinoki** [comment]
> [@Inokinoki](https://github.com/Inokinoki) I understand the explanation you give there, but it seems unexpected to me that there would be confusion between two private-linked static libraries. Especially as Python [uses the default `RTLD_LOCAL`](https://github.com/python/cpython/blob/3dfed230928de0f649061782a36691066a0ef058/Include/internal/pycore_import.h#L34-L45) when loading extension modules, so there shouldn't be any name clashes when loading two different extension modules.

Sorry for keep disturbing, I looked up the symbol `OnShutdownRun` in the attached process:

```
(lldb) image lookup -n OnShutdownRun
1 match found in /Users/*/.venv/lib/python3.10/site-packages/tink/cc/pybind/tink_bindings.cpython-310-darwin.so:
        Address: tink_bindings.cpython-310-darwin.so[0x0000000000297104] (tink_bindings.cpython-310-darwin.so.__TEXT.__text + 2700548)
        Summary: tink_bindings.cpython-310-darwin.so`google::protobuf::internal::OnShutdownRun(void (*)(void const*), void const*)
1 match found in /System/Library/PrivateFrameworks/CoreNLP.framework/Versions/A/CoreNLP:
        Address: CoreNLP[0x000000018c8935b0] (CoreNLP.__TEXT.__text + 963280)
        Summary: CoreNLP`google::protobuf::internal::OnShutdownRun(void (*)(void const*), void const*)
1 match found in /Users/*/.venv/lib/python3.10/site-packages/pyarrow/libarrow.2000.dylib:
        Address: libarrow.2000.dylib[0x0000000001b9e0f4] (libarrow.2000.dylib.__TEXT.__text + 28942580)
        Summary: libarrow.2000.dylib`google::protobuf::internal::OnShutdownRun(void (*)(void const*), void const*)
```

Does this mean that such symbol is exported several times and globally visible to all?

PS: I first `import pyarrow` and then `import tink`, where the mutex can hang infinitely.

**Comment by Inokinoki** [comment]
> Thanks for the update! Yes that's also my doubt.
> 
> 
> I will keep that in mind and try to continue debugging to figure out what
> is the case.
> […](#)

I further investigated it. And the current conclusion is similar, it's the 2 different versions of `google::protobuf::internal::OnShutdownRun` which causes the issue.

But they are good in the two libraries, because eventually they are loaded implicitly using `RTLD_LOCAL` as you said.

I think that it's because of the singleton data they are referring - `google::protobuf::internal::ShutdownData::get()::data`, which is created only once on demand:

```
auto shutdown_data = ShutdownData::get();
```

This data is expected to be different in the 2 versions (one with `absl` mutex as a member and another with `std::mutex` as a member).

I'm not sure whether it's an issue from macOS or not, but it seems that the "local" variable is shared between the versions, which caused the issue.

----------------------------------

Here are the debugging info:

So I set breakpoint on `google::protobuf::internal::OnShutdownRun` and then import `pyarrow` first. The assembly of it in `libarrow` is as follows:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.2
    frame #0: 0x000000010732a0f4 libarrow.2000.dylib`google::protobuf::internal::OnShutdownRun(void (*)(void const*), void const*)
libarrow.2000.dylib`google::protobuf::internal::OnShutdownRun:
    0x10732a0f4 <+0>:  stp    x26, x25, [sp, #-0x50]!
    0x10732a0f8 <+4>:  stp    x24, x23, [sp, #0x10]
    0x10732a0fc <+8>:  stp    x22, x21, [sp, #0x20]
    0x10732a100 <+12>: stp    x20, x19, [sp, #0x30]
    0x10732a104 <+16>: stp    x29, x30, [sp, #0x40]
    0x10732a108 <+20>: add    x29, sp, #0x40
    0x10732a10c <+24>: mov    x20, x1
    0x10732a110 <+28>: mov    x21, x0
    0x10732a114 <+32>: adrp   x8, 1794
->  0x10732a118 <+36>: ldr    x8, [x8, #0x340]
    0x10732a11c <+40>: ldaprb w8, [x8]
    0x10732a120 <+44>: adrp   x19, 1796
    0x10732a124 <+48>: ldr    x19, [x19, #0x50]
    0x10732a128 <+52>: tbz    w8, #0x0, 0x10732a238 ; <+324>
    0x10732a12c <+56>: ldr    x22, [x19]
    0x10732a130 <+60>: add    x19, x22, #0x18
    0x10732a134 <+64>: mov    x0, x19
    0x10732a138 <+68>: bl     0x107531dd0    ; symbol stub for: std::__1::mutex::lock()
    0x10732a13c <+72>: ldp    x23, x8, [x22, #0x8]
    0x10732a140 <+76>: cmp    x23, x8
    0x10732a140 <+76>: cmp    x23, x8
    0x10732a144 <+80>: b.hs   0x10732a158    ; <+100>
    0x10732a148 <+84>: stp    x21, x20, [x23]
    0x10732a14c <+88>: add    x8, x23, #0x10 
```

where the marked instruction is to get the singleton data. When I get the register bank, it contains:

```
x8 = 0x0000000107b2b8c8  guard variable for google::protobuf::internal::ShutdownData::get()::data
```

We can notice that `0x10732a138 <+68>` contains a direct call to standard C++ lib of mutex lock. And it's ok here because the singleton data also contains a `std::mutex` member.

I let it continue running and then import `tink` (with a newer version using `absl` mutex).

```
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.3
    frame #0: 0x0000000103f5b104 tink_bindings.cpython-310-darwin.so`google::protobuf::internal::OnShutdownRun(void (*)(void const*), void const*)
tink_bindings.cpython-310-darwin.so`google::protobuf::internal::OnShutdownRun:
    0x103f5b104 <+0>:  stp    x28, x27, [sp, #-0x60]!
    0x103f5b108 <+4>:  stp    x26, x25, [sp, #0x10]
    0x103f5b10c <+8>:  stp    x24, x23, [sp, #0x20]
    0x103f5b110 <+12>: stp    x22, x21, [sp, #0x30]
    0x103f5b114 <+16>: stp    x20, x19, [sp, #0x40]
    0x103f5b118 <+20>: stp    x29, x30, [sp, #0x50]
    0x103f5b11c <+24>: add    x29, sp, #0x50
    0x103f5b120 <+28>: mov    x20, x1
    0x103f5b124 <+32>: mov    x21, x0
    0x103f5b128 <+36>: adrp   x8, 501
->  0x103f5b12c <+40>: ldr    x8, [x8, #0xe8]
    0x103f5b130 <+44>: ldaprb w8, [x8]
    0x103f5b134 <+48>: adrp   x19, 502
    0x103f5b138 <+52>: ldr    x19, [x19, #0xaa0]\
    0x103f5b138 <+52>: ldr    x19, [x19, #0xaa0]
    0x103f5b13c <+56>: tbz    w8, #0x0, 0x103f5b218 ; <+276>
    0x103f5b140 <+60>: ldr    x22, [x19]
    0x103f5b144 <+64>: add    x19, x22, #0x18
    0x103f5b148 <+68>: mov    x0, x19
->  0x103f5b14c <+72>: bl     0x104026a88    ; absl::lts_20240722::Mutex::Lock()
    0x103f5b150 <+76>: ldp    x9, x8, [x22, #0x8]
    0x103f5b154 <+80>: cmp    x9, x8
    0x103f5b158 <+84>: b.hs   0x103f5b16c    ; <+104>
```

When I read the register, it's giving the same address (at the first arrow):

```
x8 = 0x00000001083278c8  guard variable for google::protobuf::internal::ShutdownData::get()::data
```

The data is already created while importing `pyarrow`, and it has a member of `std::mutex`.

But then it calls the `absl` mutex lock, which expects `absl` mutex, which can crash the program.

----------------

Hope that this helps!

**Comment by pitrou** [comment]
Thanks @Inokinoki . Unfortunately, I don't think there's anything we can do in Arrow against this. You might try to open a ticket at [protobuf](https://github.com/protocolbuffers/protobuf/), perhaps the developers there would be able to explain why that happens exactly (I don't understand why the static variable would be "shared" between the two private protobuf instances).

**Comment by pitrou** [comment]
Perhaps @bkietz understands more than me about this!

> Thanks [@Inokinoki](https://github.com/Inokinoki) . Unfortunately, I don't think there's anything we can do in Arrow against this. You might try to open a ticket at [protobuf](https://github.com/protocolbuffers/protobuf/), perhaps the developers there would be able to explain why that happens exactly (I don't understand why the static variable would be "shared" between the two private protobuf instances).

**Comment by pitrou** [comment]
Perhaps @bkietz understands more than me about this:

> I don't understand why the static variable would be "shared" between the two private protobuf instances

**Comment by maharris77** [comment]
It looks like protobuf won't fix this, and now tensorflow 2.20 has shipped with a newer protobuf that can't coexist with arrow on macOS (see also: https://github.com/tensorflow/tensorflow/issues/98563).

Since tensorflow 2.20 is the minimum version that's compatible with Python 3.13, this issue will become more important.

**Comment by pitrou** [comment]
@maharris77 I'd like to check something: what happens if you first call `sys.setdlopenflags(os.RTLD_LOCAL)` before importing PyArrow and tink?

**Edit:** you might have to make this `sys.setdlopenflags(os.RTLD_LOCAL | RTLD_NOW)`.

**Comment by pitrou** [comment]
@kou Judging by https://github.com/protocolbuffers/protobuf/issues/21686#issuecomment-2867504121, it seems like we don't hide protobuf symbols on our macOS wheels.

Similar commands on Linux give 0 results:
```console
$ nm  -C venv-3.12/lib/python3.12/site-packages/pyarrow/*.so | grep protobuf | wc -l
0
$ nm --dynamic -C venv-3.12/lib/python3.12/site-packages/pyarrow/*.so | grep protobuf | wc -l
0
```

**Comment by maharris77** [comment]
No change. I'm running into this with TF and pyarrow. It has cropped up for people across different packages; I don't have direct experience with tink or others. It's macOS specific, as explained here: https://github.com/apache/arrow/issues/33633

```python
Python 3.13.5 | packaged by conda-forge | (main, Jun 16 2025, 08:24:05) [Clang 18.1.8 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os, sys
>>> sys.setdlopenflags(os.RTLD_LOCAL | os.RTLD_NOW)
>>> import tensorflow
>>> import pyarrow  # aborts
libc++abi: terminating due to uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
zsh: abort      python
```

```python
Python 3.13.5 | packaged by conda-forge | (main, Jun 16 2025, 08:24:05) [Clang 18.1.8 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os, sys
>>> sys.setdlopenflags(os.RTLD_LOCAL | os.RTLD_NOW)
>>> import pyarrow
>>> import tensorflow  # hangs
[mutex.cc : 452] RAW: Lock blocking 0x600000968138   @

```

I also tried just `sys.setdlopenflags(os.RTLD_LOCAL)`
Similar thing happens on Python 3.12 with tensorflow 2.20, but not with tensorflow 2.19.
I can't test Python 3.13 with tensorflow 2.19 because only 2.20 is supported for 3.13.

**Comment by pitrou** [comment]
> I'm running into this with TF and pyarrow. It has cropped up for people across different packages; I don't have direct experience with tink or others.

Can you still try with tink? You just need to install it and try importing it. Just to make sure we're talking about the same problem.

**Comment by kou** [comment]
We don't use version script on macOS:

https://github.com/apache/arrow/blob/15662e67ba8742af3fdb12c48712752d578992e1/cpp/CMakeLists.txt#L361-L362

We need to find an alternative symbol visibility configuration mechanism on macOS.

**Comment by pitrou** [comment]
> We need to find an alternative symbol visibility configuration mechanism on macOS.

Apparently it's possible to use `-Wl,-exported_symbols_list`, and the symbols list file can contain wildcards (see https://stackoverflow.com/questions/6894214/how-to-create-static-library-for-ios-without-making-all-symbols-public).

**Comment by maharris77** [comment]
> Can you still try with tink? You just need to install it and try importing it. Just to make sure we're talking about the same problem.

With tink, setting flags allows me to successfully import tink and then pyarrow, but not the reverse order.

```python
Python 3.13.5 | packaged by conda-forge | (main, Jun 16 2025, 08:24:05) [Clang 18.1.8 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os, sys
>>> sys.setdlopenflags(os.RTLD_LOCAL | os.RTLD_NOW)
>>> import tink
>>> import pyarrow  # Works
>>> 
```
```python
Python 3.13.5 | packaged by conda-forge | (main, Jun 16 2025, 08:24:05) [Clang 18.1.8 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os, sys
>>> sys.setdlopenflags(os.RTLD_LOCAL | os.RTLD_NOW)
>>> import pyarrow
>>> import tink  # Hangs
[mutex.cc : 452] RAW: Lock blocking 0x6000032507f8   @
```

**Comment by maharris77** [comment]
FWIW, I made an experimental build in my fork to try addressing this. It didn’t fix things for me, but I also don't understand the build process well and just relied on Codex to suggest changes to the repo after feeding it the @kou and @pitrou comments above. Leaving it here in case it’s a useful starting point for others:

[maharris77-arrow-22.0.0.dev](https://github.com/maharris77/arrow/releases/tag/maharris77-arrow-22.0.0.dev) (commit e283e24)

Notes:
	•	I can build, make wheels, and install; the resulting pyarrow still has the same issue.
	•	To build in my environment, I also had to re-add cmake_modules to the CMAKE_MODULE_PATH in cpp/CMakeLists.txt after its removal in cpp/cmake_modules/ThirdpartyToolchain.cmake.

**Comment by miohtama** [comment]
@maharris77 Thank you for pinning this down.

I added [a StackOverflow question and answer about the matter here](https://stackoverflow.com/a/79744363/315168).

**Comment by marcelroed** [comment]
Here's another minimal reproduction of this (or a very closely related) issue:
```python
# /// script
# requires-python = ">=3.12"
# dependencies = [
#     "array-record==0.8.1",
#     "pyarrow==21.0.0",
# ]
# ///

from array_record.python import array_record_data_source
import pyarrow
```
```bash
uv run minimal_repro.py
libc++abi: terminating due to uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
```

Or, flipping the order:
```python
# /// script
# requires-python = ">=3.12"
# dependencies = [
#     "array-record==0.8.1",
#     "pyarrow==21.0.0",
# ]
# ///

import pyarrow
from array_record.python import array_record_data_source
```


```bash
uv run minimal_repro2.py
[mutex.cc : 452] RAW: Lock blocking 0x100b1eaa8   @
**HANGS FOREVER**
```

This is on MacOS 26.0.

**Comment by hartikainen** [comment]
`pyarrow==22.0` seems to have fixed this for us.

**Comment by pitrou** [comment]
> `pyarrow==22.0` seems to have fixed this for us.

Hmm, interesting. Perhaps that's because we bumped our bundled protobuf version? @raulcd 
Otherwise I'm not aware of any change on our side that might have fixed the issue.

**Comment by Inokinoki** [comment]
> > `pyarrow==22.0` seems to have fixed this for us.
> 
> Hmm, interesting. Perhaps that's because we bumped our bundled protobuf version? [@raulcd](https://github.com/raulcd) Otherwise I'm not aware of any change on our side that might have fixed the issue.

Yes, I think a newer version protobuf (which has consistent impl of absl mutex) should fix this.

> The easiest solution is to upgrade the protobuf default version of pyarrow to a later one, other than 3.21.X.

C.f. https://github.com/tink-crypto/tink-py/issues/25#issuecomment-2857960794

**Comment by pitrou** [comment]
> Yes, I think a newer version protobuf (which has consistent impl of absl mutex) should fix this.

Until the last breakage probably...

**Comment by Inokinoki** [comment]
> > Yes, I think a newer version protobuf (which has consistent impl of absl mutex) should fix this.
> 
> Until the last breakage probably...

hahaha definitely right

**Comment by ddelange** [comment]
Confirming this is fixed by pyarrow 22.0 🎉
```console
$ python -c 'from packaging.markers import default_environment; print(default_environment())'
{'implementation_name': 'cpython', 'implementation_version': '3.12.8', 'os_name': 'posix', 'platform_machine': 'arm64', 'platform_release': '24.6.0', 'platform_system': 'Darwin', 'platform_version': 'Darwin Kernel Version 24.6.0: Mon Jul 14 11:30:40 PDT 2025; root:xnu-11417.140.69~1/RELEASE_ARM64_T6041', 'python_full_version': '3.12.8', 'platform_python_implementation': 'CPython', 'python_version': '3.12', 'sys_platform': 'darwin'}
$ pip install -q tensorflow~=2.20.0 pyarrow~=21.0 && python -c 'import tensorflow; import pyarrow; print("happy")'
libc++abi: terminating due to uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
Abort trap: 6
$ pip install -q tensorflow~=2.20.0 pyarrow~=22.0 && python -c 'import tensorflow; import pyarrow; print("happy")'
happy
```

**Comment by raulcd** [comment]
From what I can see on the macOS wheels for PyArrow 22.0.0 CI jobs (I've checked both x86_64 and arm64 for a couple versions) our vcpkg dependencies are getting Protobuf 5.29.5:
See:
https://github.com/ursacomputing/crossbow/actions/runs/18656033321/job/53185254931#step:11:155
https://github.com/ursacomputing/crossbow/actions/runs/18656031313/job/53185249028#step:11:155
VCPKG upgraded from protobuf 5.29.3 to 5.29.5 some months ago but there was also a rebuild when abseil was bumped to `20250512.1`
